### PR TITLE
Plugin/Proxy - prevent nil pointer when query using a gRPC client that could not dial in

### LINKED ISSUE #1494: Plugin/Proxy - gRPC client nil pointer if the initial dial does not success
When using Proxy with gRPC, if the remote upstream does not reply when opening the communication, the 'connector' is not registered associated with the hostname.

When the DNS query get in and tries to reach the hostname, there is nothing to verify the connector exist. Which lead to the nil pointer (equivalent for go).

### PR DESCRIPTION
**Pull request by fturib**
### 1. What does this pull request do?
prevent to use a gRPC connector that does not exist.

### 2. Which issues (if any) are related?
Fix #1494 

### 3. Which documentation changes (if any) need to be made?
- no need -

### CODE REVIEW

**Review by miekg — COMMENTED** [review]
lgtm, but should probably also have a (small) test

**Comment by fturib** [comment]
right.
and looks like I pushed without a gofmt ..

File: plugin/proxy/grpc.go [hunk 1]
```diff
@@ -42,16 +43,19 @@ func (g *grpcClient) Exchange(ctx context.Context, addr string, state request.Re
...
+		}
+		d := new(dns.Msg)
+		err = d.Unpack(reply.Msg)
+		if err != nil {
+			return nil, err
+		}
+		return d, nil
 	}
-	return d, nil
+	return nil, fmt.Errorf("grpc exchange - no connexion available for host: %s ", addr)
```
Comment (grobie) [inline plugin/proxy/grpc.go]: s/connexion/connection/

Comment (fturib) [reply plugin/proxy/grpc.go]: thanks!

File: plugin/proxy/grpc.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/coredns/coredns/plugin/pkg/trace"
 	"github.com/coredns/coredns/request"
 
+	"fmt"
```
Comment (miekg) [inline plugin/proxy/grpc.go]: fmt should be moved higher

File: plugin/proxy/grpc_test.go [hunk 3]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
 	}
 }
 
+func TestRunAQuery(t *testing.T) {
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: gRPC somewhere in this test name?

File: plugin/proxy/grpc_test.go [hunk 4]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
...
+	grpclog.SetLogger(discard{})
+
+	upstream := &staticUpstream{
+		from: ".",
+		HealthCheck: healthcheck.HealthCheck{
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: forgot, but is healthcheck need for this test?

File: plugin/proxy/grpc_test.go [hunk 5]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
...
+	g := newGrpcClient(nil, upstream)
+	upstream.ex = g
+
+	p := &Proxy{}
+	p.Upstreams = &[]Upstream{upstream}
+
+	err := g.OnStartup(p)
+	if err != nil {
+		t.Errorf("Error starting grpc client exchanger: %s", err)
+		return
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: `return` is not needed, `Fatalf` is you want to quit the test.

File: plugin/proxy/grpc_test.go [hunk 6]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
...
+	}
+	// verify the client is usable, or an error is properly raised
+	state := request.Request{W: &test.ResponseWriter{}, Req: new(dns.Msg)}
+	g.Exchange(context.TODO(), "localhost:10053", state)
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: `localhost:10053` is meant to *not* answer here? Fixed port numbers are a bad idea when testing

File: plugin/proxy/grpc_test.go [hunk 7]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
...
+
+	// verify that you have proper error if the hostname is unknwn or not registered
+	_, err = g.Exchange(context.TODO(), "invalid:10055", state)
+	if err == nil {
+		t.Errorf("Expecting a proper error when querying gRPC client with invalid hostname : %s", err)
+	}
+
+	err = g.OnShutdown(p)
+	if err != nil {
+		t.Errorf("Error stopping grpc client exchanger: %s", err)
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: Fatalf

File: plugin/proxy/grpc_test.go [hunk 8]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
...
+			Hosts: []*healthcheck.UpstreamHost{
+				{
+					Name: upstreamHostname,
+				}},
+		},
+	}
+
+	filename, rmFunc, err := test.TempFile("", validCert)
+	if err != nil {
+		t.Errorf("Error saving file : %s", err)
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: Fatalf

File: plugin/proxy/grpc_test.go [hunk 9]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
...
+		return
+	}
+	defer rmFunc()
+
+	tls, err := tls.NewTLSClientConfig(filename)
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: just ignore the error here, as you control the cert's data

File: plugin/proxy/grpc_test.go [hunk 10]
```diff
@@ -59,6 +64,97 @@ func TestStartupShutdown(t *testing.T) {
...
+		return
+	}
+
+	g := newGrpcClient(tls, upstream)
+	upstream.ex = g
+
+	p := &Proxy{}
+	p.Upstreams = &[]Upstream{upstream}
+
+	// Althougth dial will not work, it is not expected to have an error
```
Comment (miekg) [inline plugin/proxy/grpc_test.go]: typo

**Review by miekg — COMMENTED** [review]
mostly nits

**Comment by fturib** [comment]
oopss .. did not see the review was done. Will address the points. Thanks!

File: plugin/proxy/grpc.go (hunk 2 above)
Comment (fturib) [reply plugin/proxy/grpc.go]: ok

File: plugin/proxy/grpc_test.go (hunk 3 above)
Comment (fturib) [reply plugin/proxy/grpc_test.go]: ok, and in the other tests

File: plugin/proxy/grpc_test.go (hunk 4 above)
Comment (fturib) [reply plugin/proxy/grpc_test.go]: no, only the pool of hosts.

File: plugin/proxy/grpc_test.go (hunk 5 above)
Comment (fturib) [reply plugin/proxy/grpc_test.go]: ok. Will also change for the others cases in same file

File: plugin/proxy/grpc_test.go (hunk 10 above)
Comment (fturib) [reply plugin/proxy/grpc_test.go]: ok

File: plugin/proxy/grpc_test.go (hunk 9 above)
Comment (fturib) [reply plugin/proxy/grpc_test.go]: ok

File: plugin/proxy/grpc_test.go (hunk 6 above)
Comment (fturib) [reply plugin/proxy/grpc_test.go]: Hum.. we need to know what is behind the address. Maybe it is better to create a Server, even though it is useless.

File: plugin/proxy/grpc_test.go [hunk 11]
```diff
@@ -14,26 +14,51 @@ import (
...
-	return []*healthcheck.UpstreamHost{
-		{
-			Name: "localhost:10053",
-		},
-		{
-			Name: "localhost:10054",
-		},
-	}
+func init() {
+	grpclog.SetLoggerV2(discardV2{})
```
Comment (fturib) [inline plugin/proxy/grpc_test.go]: SetLogger is deprecated. Changed for a V2.

File: plugin/proxy/grpc_test.go [hunk 12]
```diff
@@ -14,26 +14,51 @@ import (
...
 }
 
-func TestStartupShutdown(t *testing.T) {
-	grpclog.SetLogger(discard{})
+func buildPool(size int) ([]*healthcheck.UpstreamHost, func(), error) {
```
Comment (fturib) [inline plugin/proxy/grpc_test.go]: Build a pool of TCP Servers, to be sure that the address is used and by what.

File: plugin/proxy/grpc_test.go [hunk 13]
```diff
@@ -14,26 +14,51 @@ import (
...
+		ups[i] = &healthcheck.UpstreamHost{Name: addr}
+		srvs = append(srvs, srv)
+	}
+	stopIt := func() {
+		for _, s := range srvs {
+			s.Shutdown()
+		}
+	}
+	if len(errs) > 0 {
+		go stopIt()
```
Comment (fturib) [inline plugin/proxy/grpc_test.go]: If at lest one error when started the Servers, then stop all and return an error

File: plugin/proxy/grpc_test.go [hunk 14]
```diff
@@ -137,33 +163,38 @@ func TestRunAQueryOnSecureLinkWithInvalidCert(t *testing.T) {
...
-
-func (d discard) Fatal(args ...interface{})                 {}
-func (d discard) Fatalf(format string, args ...interface{}) {}
-func (d discard) Fatalln(args ...interface{})               {}
-func (d discard) Print(args ...interface{})                 {}
-func (d discard) Printf(format string, args ...interface{}) {}
-func (d discard) Println(args ...interface{})               {}
+type discardV2 struct{}
+
+func (d discardV2) Info(args ...interface{})                    {}
```
Comment (fturib) [inline plugin/proxy/grpc_test.go:183]: changed for a LoggerV2

File: plugin/proxy/grpc_test.go [hunk 15]
```diff
@@ -14,26 +14,51 @@ import (
...
-func pool() []*healthcheck.UpstreamHost {
-	return []*healthcheck.UpstreamHost{
-		{
-			Name: "localhost:10053",
-		},
-		{
-			Name: "localhost:10054",
-		},
-	}
+func init() {
```
Comment (fturib) [inline plugin/proxy/grpc_test.go]: the grpclog is not protected by mutex. Must be initialized safely.

**Comment by fturib** [comment]
this PR is ready for validation of review fixes.

File: plugin/proxy/grpc_test.go [hunk 16]
```diff
@@ -2,33 +2,63 @@ package proxy
...
-	"time"
 
 	"github.com/coredns/coredns/plugin/pkg/healthcheck"
 
+	"fmt"
+	"github.com/coredns/coredns/plugin/pkg/tls"
+	"github.com/coredns/coredns/plugin/test"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+	"golang.org/x/net/context"
```
Comment (miekg) [inline plugin/proxy/grpc_test.go:13]: fmt of the ordering of the impors

**Review by miekg — APPROVED** [review]

**Pull request merged**